src/coreclr/tools/aot/ILCompiler.Diagnostics/PerfMapWriter.cs

```diff
@@ -72,6 +55,96 @@ public static void Write(string perfMapFileName, int perfMapFormatVersion, IEnum
+            }
+
+            PerfmapTokensForTarget targetTokens = TranslateTargetDetailsToPerfmapConstants(details);
+
+            byte[] buffer = new byte[12];
+            if (!BitConverter.TryWriteBytes(buffer.AsSpan(0, sizeof(uint)), (uint)targetTokens.OperatingSystem)
+                || !BitConverter.TryWriteBytes(buffer.AsSpan(4, sizeof(uint)), (uint)targetTokens.Architecture)
+                || !BitConverter.TryWriteBytes(buffer.AsSpan(8, sizeof(uint)), (uint)targetTokens.Abi))
+            {
+                throw new InvalidOperationException();
+            }
+
+            if (!BitConverter.IsLittleEndian)
+            {
+                buffer.AsSpan(0, sizeof(uint)).Reverse();
+                buffer.AsSpan(4, sizeof(uint)).Reverse();
+                buffer.AsSpan(8, sizeof(uint)).Reverse();
+            }
+
+            inputHash.AddRange(buffer);
+            byte[] hash = MD5.HashData(inputHash.ToArray());
+
+            return hash;
+        }
+
+        internal record struct PerfmapTokensForTarget(PerfMapOSToken OperatingSystem, PerfMapArchitectureToken Architecture, PerfMapAbiToken Abi);
+
+        private static PerfmapTokensForTarget TranslateTargetDetailsToPerfmapConstants(TargetDetails details)
+        {
+            PerfMapOSToken osToken = details.OperatingSystem switch
+            {
+                TargetOS.Unknown => PerfMapOSToken.Unknown,
+                TargetOS.Windows => PerfMapOSToken.Windows,
+                TargetOS.Linux => PerfMapOSToken.Linux,
+                TargetOS.OSX => PerfMapOSToken.OSX,
+                TargetOS.FreeBSD => PerfMapOSToken.FreeBSD,
+                TargetOS.NetBSD => PerfMapOSToken.NetBSD,
+                TargetOS.SunOS => PerfMapOSToken.SunOS,
+                TargetOS.WebAssembly => PerfMapOSToken.WebAssembly,
+                _ => throw new NotImplementedException(details.OperatingSystem.ToString())
+            };
+
+            PerfMapAbiToken abiToken = details.Abi switch
+            {
+                TargetAbi.Unknown => PerfMapAbiToken.Unknown,
+                TargetAbi.CoreRT => PerfMapAbiToken.CoreRT,
+                TargetAbi.CoreRTArmel => PerfMapAbiToken.CoreRTArmel,
+                TargetAbi.Jit => PerfMapAbiToken.Jit,
+                TargetAbi.CppCodegen => PerfMapAbiToken.CppCodegen,
+                _ => throw new NotImplementedException(details.Abi.ToString())
+            };
+
+            PerfMapArchitectureToken archToken = details.Architecture switch
+            {
+                TargetArchitecture.Unknown => PerfMapArchitectureToken.Unknown,
+                TargetArchitecture.ARM => PerfMapArchitectureToken.ARM,
+                TargetArchitecture.ARM64 => PerfMapArchitectureToken.ARM64,
+                TargetArchitecture.X64 => PerfMapArchitectureToken.X64,
+                TargetArchitecture.X86 => PerfMapArchitectureToken.X86,
+                TargetArchitecture.Wasm32 => PerfMapArchitectureToken.Wasm32,
```

Comment:
I would leave Wasm out of this until we decided whether a R2R perfmap for wasm even makes sense, which is pretty far out.